luanti-org/luanti
# Visual Effects Vol. 2

This PR seeks to improve the visuals of Luanti through improvements to existing effects and addition of new ones (pertains to point 2.1 of the roadmap). To this end, shader code, and to a lesser extent engine code is modified. 
<details>
<summary>List of changes</summary>

### Adjustments to effects from #14610 and others:
- Translucent foliage sees some improvements to consistency and reduced artifacts
- Water reflections now work without waving water enabled
- Water reflections see some changes to correct for low alpha
- The (still placeholder) specular maps are improved
- The minimap shader is pixelated
- Made depth attenuation in volumetric lighting optional
- Made tinted fog optional

### New effects:
- Added gamma setting
- Vignette as part of the second stage shaders
- [ASL CDL](https://en.wikipedia.org/wiki/ASC_CDL) color grading
- Tinted sunlight
- Adjustable artificial light color (global, not per node :( )

### Lua API additions
- `lighting.vignette` controls the second stage vignette.
    - `dark` how bright the darkest part of the vignette is, default: `0.3`.
    - `bright` how bright the brightest part of the vignette is, default: `1.1`.
    - `power` controls the blending between dark and bright, default: `1.1`.
- `lighting.cdl` controls the color grading decision list.
    - The output color follows the formula `out = pow(in * slope + offset, power)`.
    - Default: `{ slope = { x = 1.2, y = 1.0, z = 0.8}, offset = { x = 0.0, y = 0.0, z = 0.0 }, power = { x = 1.25, y = 1.0, z = 0.9 } }`.
- `lighting.volumetric_light.beta_r0` controls the absorbance of the atmosphere for volumetric light and tinted sunlight effects. 
    - Default: `{ x = 3.336e-01, y = 8.754e-01, z = 1.953 }`. 
    - The default is reasonable for most purposes, but this can be used to create sunrises/sunsets of arbitrary color.
    - An absorbance of zero `{ x = 0, y = 0, z = 0 }` is equivalent to disabling tinted sunlight, except it also applies to volumetrics.
- `lighting.artificial_light` Controls the color of artificial light, default: `{ r = 133, g = 133, b = 133 }`.
- `lighting.foliage_translucency` controls the strength of the translucent foliage effect, default: `1.5`.
- `lighting.specular_intensity` controls the strength of the specular reflection effect, default: `1.5`.

### Settings
| Name | Technical Name | Location | Type | Default |
|---|---|---|---|---|
| Tinted sunlight | enable_sun_tint | Effects>Dynamic Shadows | bool | false |
| Gamma | secondstage_gamma | Effects>Post Processing | float | 1.6 |
| Color grading | enable_color_grading | Effects>Post Processing | bool | false |
| Volumetric Depth Attenuation | enable_volumetric_depth_attenuation | Effects>Post Processing | bool | false |
| Tinted fog | enable_tinted_fog | Effects>Other Effects | bool | false |
| Bump maps | enable_bumpmaps | Effects>Other Effects | bool | false |
</details>

<details>
<summary>Screenshots</summary>

![screenshot_20241027_211038](https://github.com/user-attachments/assets/c9d7ca83-f8e7-4fd0-accd-6f197302f963)
![screenshot_20241027_214630](https://github.com/user-attachments/assets/3ffc771a-6f66-4bcb-a19f-e61789eab4bd)
![screenshot_20241110_181206](https://github.com/user-attachments/assets/1fa93cf5-9669-4968-bf7b-5f29b9747a13)
![screenshot_20241110_181749](https://github.com/user-attachments/assets/c5c3b903-7c89-4ce1-8efd-564344eb3bf5)
![screenshot_20241110_191126](https://github.com/user-attachments/assets/70c226fd-279b-4efd-84ff-0d0c5ec25a14)
</details>

## To do
This PR is ready for review.

## How to test
For most features, simply change the game settings. Of course, the game you use should support dynamic shadows and volumetrics. An example for the Lua controls is shown below:
```Lua
player:set_lighting({
    artificial_light = { r = 140, g = 120, b = 100 },
    cdl = { offset = { x = 1, y = 1, z = 1 }, slope = { x = -1, y = -1, z = -1 }, power = { x = 2, y = 2, z = 2 } },
    vignette = { dark = 0, bright = 2, power = 0.5 },
    volumetric_light = { beta_r0 = { x = 2, y = 1, z = 0.5 } },
    foliage_translucency = 4.0, 
    specular_intensity = 4.0
})
```
This should produce inverted colors with color grading enabled, an exaggerated vignette, very bright reflections and foliage, blue sunsets and yellow artificial light.
